aidial_adapter_bedrock/llm/converse/adapter.py

```diff
@@ -0,0 +1,174 @@
+        Callable[[List[Tuple[ConverseMessage, Any]]], Awaitable[int]],
+    ]
+    support_tools: bool
+    partitioner: Callable[[List[Tuple[ConverseMessage, Any]]], List[int]] = (
+        turn_based_partitioner
+    )
+
+    async def _discard_messages(
+        self, params: ConverseParams, max_prompt_tokens: int | None
+    ) -> Tuple[DiscardedMessages | None, ConverseParams]:
+        if max_prompt_tokens is None:
+            return None, params
+
+        discarded_messages, messages = await truncate_prompt(
+            messages=params.messages.list,
+            tokenizer=self.input_tokenizer_factory(self.deployment, params),
+            keep_message=keep_last,
+            partitioner=self.partitioner,
+            model_limit=None,
+            user_limit=max_prompt_tokens,
+        )
+
+        return list(
+            params.messages.to_original_indices(discarded_messages)
+        ), ConverseParams(
+            **{
+                **params.to_dict(),
+                "messages": ListProjection(
+                    omit_by_indices(messages, discarded_messages)
+                ),
+            },
+        )
+
+    async def count_prompt_tokens(
+        self, params: ModelParameters, messages: List[DialMessage]
+    ) -> int:
+        converse_params = await self.construct_converse_params(messages, params)
+        return await self.input_tokenizer_factory(
+            self.deployment, converse_params
+        )(converse_params.messages.list)
+
+    async def count_completion_tokens(self, string: str) -> int:
+        return self.tokenize_text(string)
+
+    async def compute_discarded_messages(
+        self, params: ModelParameters, messages: List[DialMessage]
+    ) -> DiscardedMessages | None:
+        converse_params = await self.construct_converse_params(messages, params)
+        discarded_messages, _ = await self._discard_messages(
+            converse_params, params.max_prompt_tokens
+        )
+        return discarded_messages
+
+    def get_tool_config(self, params: ModelParameters) -> ConverseTools | None:
+        if params.tool_config and not self.support_tools:
+            raise ValidationError("Tools are not supported")
+        return (
+            to_converse_tools(params.tool_config)
+            if params.tool_config
+            else None
+        )
+
+    async def construct_converse_params(
+        self,
+        messages: List[DialMessage],
+        params: ModelParameters,
+    ) -> ConverseParams:
+        system_message = get_converse_system_prompt(messages)
+        processed_messages = await process_messages(messages, self.storage)
+        if not processed_messages.list:
+            raise ValidationError("List of messages must not be empty")
+
+        return ConverseParams(
+            system=[system_message] if system_message else None,
+            messages=processed_messages,
+            inferenceConfig=cast(
+                InferenceConfig,
```

Comment:
casting should be redundant

Reply:
It's not redundant, since `remove_none` messes typing

I could make something like that:
```python

T = TypeVar("T", bound=dict)


def remove_nones(d: T) -> T:
    return cast(T, {k: v for k, v in d.items() if v is not None})

```
But that's much more tricky, since `T` could have `class AllPresentKeys(TypedDict, total=True`)...`

Reply:
do you really need `remove_nones` then?

Reply:
Yes, I do
Converse API doesn't allow null value for such keys:

https://docs.aws.amazon.com/bedrock/latest/APIReference/API_runtime_InferenceConfiguration.html

Reply:
Replaced `remove_nones` with 
```python
inferenceConfig=InferenceConfig(
                **{
                    key: value
                    for key, value in [
                        ("temperature", params.temperature),
                        ("topP", params.top_p),
                        ("maxTokens", params.max_tokens),
                        ("stopSequences", params.stop),
                    ]
                    if value is not None
                }
            ),
```

Reply:
But basically, it's the same trick as `cast`, since I just rely on the fact, that linter has limited understanding of unpacking